packages/nodejs/src/scope.ts

```diff
@@ -221,30 +227,85 @@ export class ScopeManager {
     // prevent re-wrapping
     contextWrapper[WRAPPED] = true
 
-    // explicitly inherit the original function's length, because it is
-    // otherwise zero-ed out
+    // explicitly inherit the original function's length and name
     Object.defineProperty(contextWrapper, "length", {
-      enumerable: false,
       configurable: true,
-      writable: false,
       value: fn.length
     })
 
+    Object.defineProperty(contextWrapper, "name", {
+      configurable: true,
+      value: fn.name
+    })
+
     return contextWrapper
   }
 
   public emitWithContext(ee: EventEmitter): void {
     // eslint-disable-next-line @typescript-eslint/no-this-alias
-    const that = this
+    const thisScopeManager = this
+
+    // For each event and callback that is registered with the event emitter,
+    // `boundCallbacks` keeps a reference to the newly-created context-bound
+    // callback that wraps around the original callback.
+    // A listener with the same event and callback could be added from
+    // different asynchronous contexts, meaning its context-bound wrapper
+    // should bind it to a different scope, so it keeps a list of
+    // context-bound callbacks for a single event and callback pair.
+    const boundCallbacks = new BoundCallbacks()
 
-    EVENT_EMITTER_METHODS.forEach(method => {
+    EVENT_EMITTER_ADD_METHODS.forEach(method => {
       if (ee[method]) {
         shimmer.wrap(ee, method, (oldFn: Func<any>) => {
           return function (this: unknown, event: string, cb: Func<void>) {
-            return oldFn.call(this, event, that.bindContext(cb))
+            const boundCallback = thisScopeManager.bindContext(cb)
+            boundCallbacks.push(event, cb, boundCallback)
+            return oldFn.call(this, event, boundCallback)
           }
         })
       }
     })
+
+    EVENT_EMITTER_REMOVE_METHODS.forEach(method => {
+      if (ee[method]) {
+        shimmer.wrap(ee, method, (oldFn: Func<any>) => {
+          return function (this: unknown, event: string, cb: Func<void>) {
+            // If there is no bound callback for this event and callback, it
+            // might be a listener that was added before the event emitter was
+            // wrapped, so we should attempt to remove the given callback.
+            const maybeBoundCallback = boundCallbacks.pop(event, cb) ?? cb
+            return oldFn.call(this, event, maybeBoundCallback)
+          }
+        })
+      }
+    })
+  }
+}
+
+class BoundCallbacks extends Map<string, WeakMap<Func<void>, Func<void>[]>> {
```

Comment:
A bit of a nitpick: With this class extending Map it also exposes a lot of other public functions, but it looks like only `push` and `pop` should really be used. Functions that aren't part of the Map API, e.g. `new BoundCallbacks().get("foo")`. Should we instead create a new class that wraps around the `Map` and only exposes the `push` and `pop` functions?

```class
class BoundCallbacks {
  #map: Map<...>

  push(...) { this.#map.set(...) }
  pop(...) { this.#map.get(...) }
}
```

Reply:
Thanks! You're right, it shouldn't really expose methods other than `push` and `pop`.